apps/build-onchain-apps/contracts/src/SignatureMintERC721.sol

```diff
@@ -0,0 +1,143 @@
+ */
+contract SignatureMintERC721 is ERC721 {
+    using ECDSA for bytes32;
+    using MessageHashUtils for bytes32;
+
+    // Contract Params
+    uint256 public paidMintCost = 0.0001 ether;
+    address public freeMintSigner; // Wallet Address for the authorized wallet to distribute free mints.
+    uint256 public currentId = 0; // Current NFT id
+    mapping(address => uint256) public usedFreeMints; // Mapping to store free mints
+
+    // Setup contract Metadata
+    string public imageUri = "ipfs://QmQRzw5MtNQcdjg9rZ8MYugAz6WjaBdvsuNShKTLYNJdj7/coffee_dog.png";
+    NFTUtilities.CollectionMetadata public metadata =
+        NFTUtilities.CollectionMetadata("CoffeeDog", "Buy a dog a coffee", "CDOG", imageUri, "");
+
+    // Events
+    event Mint(address indexed buyer, uint256 tokenId, uint256 cost);
+
+    constructor(address signer) ERC721(metadata.name, metadata.symbol) {
+        freeMintSigner = signer;
+    }
+
+    /**
+     * @dev Function to mint for free using a wallet signature
+     * @param  account that will receive the NFT
+     * @param  signature that authorizes this transaction to mint for free.
+     */
+    function freeMint(address account, bytes memory signature) external {
+        require(this.verifySignature(account, signature) == true, "Unable to verify Signature");
+
+        // Check if the free mint has already been used (1 per wallet)
+        require(usedFreeMints[account] == 0, "Free mint already used");
+        usedFreeMints[account] = 1;
+
+        // Mint the NFT
+        _mint(account, ++currentId);
+        emit Mint(account, currentId, 0);
+    }
+
+    /**
+     * @dev Function to mint for a fixed cost defined in the paidMintCost parameter
+     * @param  account that will receive the NFT
+     */
+    function mint(address account) external payable {
+        if (msg.value != paidMintCost) {
+            revert InsufficientFunds();
+        }
+        _mint(account, ++currentId);
+        emit Mint(account, currentId, paidMintCost);
+    }
+
+    /**
+     * @dev Function to return the contract metadata, this is stored on contract versus IPFS.  This ensures
+     *      the collection information renders correctly on NFT marketplaces.
+     */
+    function contractURI() public view returns (string memory) {
+        bytes memory jsonMetadata = abi.encodePacked(
+            "{",
+            NFTUtilities.getEncodedStringPair("description", metadata.description, false),
+            NFTUtilities.getEncodedStringPair("symbol", metadata.symbol, false),
+            NFTUtilities.getEncodedStringPair("image", metadata.image, false),
+            NFTUtilities.getEncodedStringPair("animation_url", metadata.animation_url, true),
+            "}"
+        );
+        return string(jsonMetadata);
+    }
+
+    function tokenURI(uint256 tokenId) public view override(ERC721) returns (string memory) {
+        require(tokenId <= currentId, "Token doesnt exist");
```

Comment:
Weird condition where this check passes for tokenId 0 on creation. Maybe should be nextTokenID and strictly less than?